pulsar-client/src/main/java/org/apache/pulsar/client/impl/ClientCnx.java

```diff
@@ -187,13 +194,32 @@ public void channelActive(ChannelHandlerContext ctx) throws Exception {
                 });
     }
 
-    protected ByteBuf newConnectCommand() throws PulsarClientException {
-        String authData = "";
-        if (authentication.getAuthData().hasDataFromCommand()) {
+    private boolean isMutualAuthenticationMethod() {
+        return false;
+    }
+
+    protected ByteBuf newConnectCommand() throws IOException {
+        // mutual authentication is to auth between `remoteHostName` and this client for this channel.
+        // each channel will have a mutual client/server pair, mutual client evaluateChallenge with init data,
+        // and return authData to server.
+        if (isMutualAuthenticationMethod()) {
```

Comment:
I think there are some commonalities between these branches, we should try to consolidate into one interface and hide the implementation. If I understand this correctly:

- isMutalAuthenticationMethod == true => authData = getAuthData(remoteHostName).authenticate("init".getBytes());
- hasDataFromCommand => authData = = getAuthData().getCommandData();
- otherwise => authData = ""

So I think the Authentication can just provide a method `byte[] authenticate(String remoteHostname)`? Hide the implementation details to individual provider?